homeassistant/components/media_player/intent.py

```diff
@@ -19,13 +25,44 @@
 INTENT_MEDIA_NEXT = "HassMediaNext"
 INTENT_SET_VOLUME = "HassSetVolume"
 
-DATA_LAST_PAUSED = f"{DOMAIN}.last_paused"
+_PAUSE_DELAY: Final = 5
+
+
+@dataclass
+class LastPaused:
+    """Information about last media players that were paused by voice."""
+
+    timestamp: float | None = None
+    entity_ids: set[str] = field(default_factory=set)
+    waiting_entity_ids: set[str] = field(default_factory=set)
 
 
 async def async_setup_intents(hass: HomeAssistant) -> None:
     """Set up the media_player intents."""
-    intent.async_register(hass, MediaUnpauseHandler())
-    intent.async_register(hass, MediaPauseHandler())
+    last_paused = LastPaused()
+
+    def state_changed_listener(event: Event[EventStateChangedData]) -> None:
+        """Update last paused timestamp as expected media players enter their paused states."""
+        if (
+            ((state := event.data.get("new_state")) is None)
+            or (state.domain != DOMAIN)
+            or (state.state != STATE_PAUSED)
+        ):
+            return
+
+        entity_id = state.entity_id
+        if entity_id not in last_paused.waiting_entity_ids:
+            return
+
+        # Move out of waiting and bump timestamp
+        last_paused.waiting_entity_ids.remove(entity_id)
+        last_paused.entity_ids.add(entity_id)
+        last_paused.timestamp = time.time()
+
+    hass.bus.async_listen(EVENT_STATE_CHANGED, state_changed_listener)
```

Comment:
Can we do it without listening to every state change in Home Assistant? Why can't we compare timestamps ?

Reply:
At least, let your data class track only when it has data. And stop listening when it expires. But preferably we never listen to it.

Reply:
Removed in favor of using context